flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamGraphGeneratorExecutionModeDetectionTest.java

```diff
@@ -19,37 +19,129 @@
 import org.apache.flink.streaming.api.environment.CheckpointConfig;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
 import org.apache.flink.streaming.api.operators.SimpleOperatorFactory;
 import org.apache.flink.streaming.api.operators.SourceOperatorFactory;
 import org.apache.flink.streaming.api.transformations.SourceTransformation;
 import org.apache.flink.streaming.api.transformations.TwoInputTransformation;
 import org.apache.flink.util.TestLogger;
 
+import org.hamcrest.Description;
+import org.hamcrest.TypeSafeMatcher;
 import org.junit.Test;
 
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 
 import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertThat;
 
 /**
  * Tests for the detection of the {@link RuntimeExecutionMode runtime execution mode} during
  * stream graph translation.
  */
 public class StreamGraphGeneratorExecutionModeDetectionTest extends TestLogger {
 
+	@Test
+	public void testExecutionModePropagationFromEnvWithDefaultAndBoundedSource() {
+		final StreamExecutionEnvironment environment =
+				StreamExecutionEnvironment.getExecutionEnvironment();
+
+		environment
+				.fromSource(
+						new MockSource(Boundedness.BOUNDED, 100),
+						WatermarkStrategy.noWatermarks(),
+						"bounded-source")
+				.print();
+
+		assertThat(
+				environment.getStreamGraph(),
+				hasProperties(
+						GlobalDataExchangeMode.ALL_EDGES_PIPELINED,
+						ScheduleMode.EAGER,
+						true));
+	}
+
+	@Test
+	public void testExecutionModePropagationFromEnvWithDefaultAndUnboundedSource() {
+		final StreamExecutionEnvironment environment =
+				StreamExecutionEnvironment.getExecutionEnvironment();
+
+		environment
+				.fromSource(
+						new MockSource(Boundedness.CONTINUOUS_UNBOUNDED, 100),
+						WatermarkStrategy.noWatermarks(),
+						"unbounded-source")
+				.print();
+
+		assertThat(
+				environment.getStreamGraph(),
+				hasProperties(
+						GlobalDataExchangeMode.ALL_EDGES_PIPELINED,
+						ScheduleMode.EAGER,
+						true));
+	}
+
+	@Test
+	public void testExecutionModePropagationFromEnvWithAutomaticAndBoundedSource() {
+		final Configuration config = new Configuration();
+		config.set(ExecutionOptions.RUNTIME_MODE, RuntimeExecutionMode.AUTOMATIC);
+
+		final StreamExecutionEnvironment environment =
+				StreamExecutionEnvironment.getExecutionEnvironment();
+		environment.configure(config, getClass().getClassLoader());
+
+		environment
+				.fromSource(
+						new MockSource(Boundedness.BOUNDED, 100),
+						WatermarkStrategy.noWatermarks(),
+						"bounded-source")
+				.print();
+
+		assertThat(
+				environment.getStreamGraph(),
+				hasProperties(
+						GlobalDataExchangeMode.POINTWISE_EDGES_PIPELINED,
+						ScheduleMode.LAZY_FROM_SOURCES_WITH_BATCH_SLOT_REQUEST,
+						false));
+	}
+
+	@Test(expected = IllegalStateException.class)
```

Comment:
Maybe using the `ExceptionException` rule would be better here because it's more fine grained and allows matching on the exception message.